File: nodes/core/market/indicators/atrx_indicator_node.py
import logging
from typing import Any, cast

import numpy as np
import pandas as pd

from core.types_registry import AssetSymbol, IndicatorResult, IndicatorType, IndicatorValue, get_type
from nodes.core.market.indicators.base.base_indicator_node import BaseIndicator
from services.indicator_calculators.atrx_calculator import calculate_atrx

logger = logging.getLogger(__name__)


class AtrXIndicator(BaseIndicator):
    """
    Computes the ATRX indicator for a single asset's OHLCV data.
    A = ATR% = ATR / Last Done Price
    B = % Gain From 50-MA = (Price - SMA50) / SMA50
    ATRX = B / A = (% Gain From 50-MA) / ATR%

    Reference:
        https://www.tradingview.com/script/oimVgV7e-ATR-multiple-from-50-MA/
    """

    inputs = {"ohlcv": get_type("OHLCVBundle")}
    outputs = {"results": list[IndicatorResult]}
    default_params = {
        "length": 14,  # ATR period
        "ma_length": 50,  # SMA period for trend calculation
    }
    params_meta = [
        {"name": "length", "type": "integer", "default": 14, "description": "ATR period"},
        {
            "name": "ma_length",
            "type": "integer",
            "default": 50,
            "description": "SMA period for trend calculation",
        },
    ]

    def _map_to_indicator_value(
        self, ind_type: IndicatorType, raw: dict[str, Any]
    ) -> IndicatorValue:
        """
        Satisfy BaseIndicatorNode's abstract contract. ATRX node uses its own
        _execute_impl path and does not rely on base mapping.
        """
        return IndicatorValue(single=float("nan"))

    async def _execute_impl(self, inputs: dict[str, Any]) -> dict[str, Any]:
        ohlcv_bundle: dict[AssetSymbol, list[dict[str, float]]] = inputs.get("ohlcv", {})
        if not ohlcv_bundle:
            logger.warning("Empty OHLCV bundle provided to ATRX indicator")
            return {"results": []}

        # Get the first (and typically only) symbol's bars
        ohlcv = next(iter(ohlcv_bundle.values()))
        
        if not ohlcv:
            logger.warning("Empty OHLCV data provided to ATRX indicator")
            return {"results": []}

        try:
            # Check for minimum data requirements first
            length_value = int(cast(int, self.params.get("length", 14)))
            ma_length_value = int(cast(int, self.params.get("ma_length", 50)))
            min_required = max(length_value, ma_length_value)
            if len(ohlcv) < min_required:
                logger.warning(
                    f"Insufficient data for ATRX calculation: {len(ohlcv)} bars, need {min_required}"
                )
                return {"results": []}

            # Extract lists directly from OHLCV data
            high_prices = [bar["high"] for bar in ohlcv]
            low_prices = [bar["low"] for bar in ohlcv]
            close_prices = [bar["close"] for bar in ohlcv]

            # Create minimal DataFrame for calculator API
            df_calc = pd.DataFrame(
                {
                    "high": high_prices,
                    "low": low_prices,
                    "close": close_prices,
                }
            )

            # Call calculator directly
            atrx_result = calculate_atrx(
                df_calc,
                length=length_value,
                ma_length=ma_length_value,
                source="close",
            )
            atrx_values = atrx_result.get("atrx", [])

            if not atrx_values:
                logger.warning("ATRX calculation returned empty results")
                return {"results": []}

            # Get the last value
            atrx_value = atrx_values[-1]

            # Filter out NaN results (e.g., zero volatility cases)
            if atrx_value is None or np.isnan(atrx_value):
                logger.warning("ATRX calculation resulted in NaN (likely zero volatility)")
                return {"results": []}

            result = IndicatorResult(
                indicator_type=IndicatorType.ATRX,
                timestamp=int(ohlcv[-1]["timestamp"]),
                values=IndicatorValue(single=atrx_value),
            )
            return {"results": [result]}

        except Exception as e:
            logger.error(f"Error calculating ATRX indicator: {e}")
            return {"results": []}

File: packages/core/src/nodes/core/market/indicators/atrx-indicator-node.ts
// src/nodes/core/market/indicators/atrx-indicator-node.ts
// Translated from: nodes/core/market/indicators/atrx_indicator_node.py

import { BaseIndicator } from './base/base-indicator-node';
import {
  IndicatorType,
  createIndicatorResult,
  createIndicatorValue,
  getType,
} from '../../../../core/types';
import type {
  ParamMeta,
  DefaultParams,
  NodeInputs,
  NodeOutputs,
  IndicatorValue,
  OHLCVBar,
  OHLCVBundle,
  NodeUIConfig,
} from '../../../../core/types';
import { calculateAtrx } from '../../../../services/indicator-calculators';

/**
 * Computes the ATRX indicator for a single asset's OHLCV data.
 * A = ATR% = ATR / Last Done Price
 * B = % Gain From 50-MA = (Price - SMA50) / SMA50
 * ATRX = B / A = (% Gain From 50-MA) / ATR%
 *
 * Reference:
 *     https://www.tradingview.com/script/oimVgV7e-ATR-multiple-from-50-MA/
 */
export class AtrXIndicator extends BaseIndicator {
  static uiConfig: NodeUIConfig = {
    size: [220, 100],
    displayResults: false,
    resizable: false,
  };

  static override inputs: Record<string, unknown> = {
    ohlcv: getType('OHLCVBundle'),
  };

  static override outputs: Record<string, unknown> = {
    results: Array, // list[IndicatorResult]
  };

  static override defaultParams: DefaultParams = {
    length: 14, // ATR period
    ma_length: 50, // SMA period for trend calculation
  };

  static override paramsMeta: ParamMeta[] = [
    { name: 'length', type: 'integer', default: 14, description: 'ATR period' },
    {
      name: 'ma_length',
      type: 'integer',
      default: 50,
      description: 'SMA period for trend calculation',
    },
  ];

  protected mapToIndicatorValue(
    _indType: IndicatorType,
    _raw: Record<string, unknown>
  ): IndicatorValue {
    // ATRX node uses its own _executeImpl path and does not rely on base mapping.
    return createIndicatorValue({ single: NaN });
  }

  protected override async executeImpl(inputs: NodeInputs): Promise<NodeOutputs> {
    const ohlcvBundle = inputs.ohlcv as OHLCVBundle | undefined;
    if (!ohlcvBundle || ohlcvBundle.size === 0) {
      console.warn('Empty OHLCV bundle provided to ATRX indicator');
      return { results: [] };
    }

    // Get the first (and typically only) symbol's bars
    const ohlcv = ohlcvBundle.values().next().value as OHLCVBar[] | undefined;

    if (!ohlcv || ohlcv.length === 0) {
      console.warn('Empty OHLCV data provided to ATRX indicator');
      return { results: [] };
    }

    try {
      // Check for minimum data requirements first
      const lengthRaw = this.params.length ?? 14;
      const maLengthRaw = this.params.ma_length ?? 50;

      const lengthValue = typeof lengthRaw === 'number' ? Math.floor(lengthRaw) : 14;
      const maLengthValue = typeof maLengthRaw === 'number' ? Math.floor(maLengthRaw) : 50;

      const minRequired = Math.max(lengthValue, maLengthValue);
      if (ohlcv.length < minRequired) {
        console.warn(
          `Insufficient data for ATRX calculation: ${ohlcv.length} bars, need ${minRequired}`
        );
        return { results: [] };
      }

      // Extract lists directly from OHLCV data
      const highPrices = ohlcv.map((bar) => bar.high);
      const lowPrices = ohlcv.map((bar) => bar.low);
      const closePrices = ohlcv.map((bar) => bar.close);

      // Call calculator directly with required parameters
      const atrxResult = calculateAtrx(
        highPrices,
        lowPrices,
        closePrices,
        closePrices, // prices = close
        lengthValue,
        maLengthValue
      );
      const atrxValues = atrxResult.atrx;

      if (!atrxValues || atrxValues.length === 0) {
        console.warn('ATRX calculation returned empty results');
        return { results: [] };
      }

      // Get the last value
      const atrxValue = atrxValues[atrxValues.length - 1];

      // Filter out NaN results (e.g., zero volatility cases)
      if (atrxValue === null || Number.isNaN(atrxValue)) {
        console.warn('ATRX calculation resulted in NaN (likely zero volatility)');
        return { results: [] };
      }

      const result = createIndicatorResult({
        indicatorType: IndicatorType.ATRX,
        timestamp: ohlcv[ohlcv.length - 1].timestamp,
        values: createIndicatorValue({ single: atrxValue }),
      });

      return { results: [result] };
    } catch (error) {
      console.error('Error calculating ATRX indicator:', error);
      return { results: [] };
    }
  }
}
